### ISSUE TITLE
jsonb arrays are converted incorrectly

### ISSUE THREAD

**Issue by felixge** [plv8/plv8js#356]
plv8 3.0alpha seems to have a bug when it comes to converting arrays of objects from jsonb to v8 arrays. All except for the last element of the array seem to be dropped during the conversion.

Steps to reproduce:

```
psql (11.2)
Type "help" for help.

plv8_test=# CREATE EXTENSION plv8;
CREATE EXTENSION
plv8_test=# SELECT plv8_version();
 plv8_version
--------------
 3.0alpha
(1 row)

plv8_test=# CREATE OR REPLACE FUNCTION jsonb_cat(data jsonb) RETURNS jsonb LANGUAGE plv8 AS
plv8_test-# $$
plv8_test$#     return data;
plv8_test$# $$;
CREATE FUNCTION
plv8_test=# SELECT jsonb_cat('[{"a": 1},{"b": 2},{"c": 3}]'::jsonb);
 jsonb_cat
------------
 [{"c": 3}]
(1 row)
```

Expected result:

```
[{"a": 1},{"b": 2},{"c": 3}]
```

FWIW this bug does not occur when using the `json` type rather than `jsonb`. Version `2.3.12` also works correctly, so this seems to be a regression in `3.0alpha`.

Please let me know if I can provide more information.

**Comment by JerrySievert** [comment]
it has to do with the direct jsonb conversion - thanks for supplying a test case!

I'll try to get it fixed in the next day or so.

you can disable this and use the old functionality by commenting out line 55 in Makefile if you need it addressed immediately:

```
CCFLAGS += -DJSONB_DIRECT_CONVERSION
```

**Comment by felixge** [comment]
@JerrySievert thanks for the quick reply! There is no rush on this from my end, I'm okay with using an older release for now.

**Comment by JerrySievert** [comment]
*stupid* error found and fixed, doing a little more then I'll open and merge a PR.

**Comment by felixge** [comment]
@JerrySievert awesome, thank you for looking into this : ). Btw. is this `DJSONB_DIRECT_CONVERSION` aimed at making it faster to pass large jsonb blobs to plv8 functions? If yes, that'd be awesome : ).

Unrelated: I have an idea for a feature to add to plv8 and would love to hear your thoughts. Should I open a new issue for it? I've already got a small proof-of-concept patch working, but I've decided to level up my C/C++ skills a little more before continuing with it.

**Comment by JerrySievert** [comment]
yes, `JSONB_DIRECT_CONVERSION` makes it faster and more memory-friendly when dealing with `JSONB`.

when `JSONB` was first added to pg, the conversion was to do an internal cast to `JSON` and use `JSON.parse()` (the c++ version, not the javascript version) to bring it in - that ended up with an extra copy of it in memory.

it still does a full conversion, since you're passing it in to javascript - I suppose that the  most efficient way would be to intercept the properties and do the conversion on access, but I'm not quite there yet.

feel free to open another issue, or email me, whichever works for you - today is my "plv8 day", so I'll be thinking about plv8 all day :)

**Comment by JerrySievert** [comment]
live on 3.0alpha and back ported to 2.3.13 - closing.

**Comment by felixge** [comment]
@JerrySievert thank you so much! I'll give it a shot very soon!

**Comment by cliang57** [comment]
I found the same problem for PLV8 2.3.11 with JSONB_DIRECT_CONVERSION on. Is this fixed in the latest release like 2.3.13?

**Comment by JerrySievert** [comment]
as noted above:

> live on 3.0alpha and back ported to 2.3.13 - closing.
